File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/DexService.java
package sk.styk.martin.apkanalyzer.business.service;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.support.annotation.NonNull;
import android.util.Log;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;

import dalvik.system.DexFile;
import sk.styk.martin.apkanalyzer.model.detail.ClassPathData;

/**
 * Created by Martin Styk on 21.10.2017.
 */
public class DexService {

    public ClassPathData get(@NonNull PackageInfo packageInfo) {

        List<String> packageClasses = new ArrayList<>();
        List<String> otherClasses = new ArrayList<>();
        int innerClasses = 0;
        ClassPathData classPathData = new ClassPathData();

        ApplicationInfo applicationInfo = packageInfo.applicationInfo;

        if (applicationInfo != null) {
            String packageName = applicationInfo.packageName;
            DexFile dexFile = null;
            try {
                dexFile = new DexFile(applicationInfo.sourceDir);
                for (Enumeration<String> iterator = dexFile.entries(); iterator.hasMoreElements(); ) {
                    String className = iterator.nextElement();
                    if (className != null && className.startsWith(packageName))
                        packageClasses.add(className);
                    else
                        otherClasses.add(className);

                    if (className != null && className.contains("$")) {
                        innerClasses++;
                    }
                }
            } catch (IOException e) {
                Log.e(DexService.class.getSimpleName(), e.getLocalizedMessage());
            } finally {
                if (dexFile != null) {
                    try {
                        dexFile.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        }

        classPathData.setPackageClasses(packageClasses);
        classPathData.setOtherClasses(otherClasses);
        classPathData.setNumberOfInnerClasses(innerClasses);

        return classPathData;
    }
}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/analysis/logic/DexService.kt
package sk.styk.martin.apkanalyzer.business.analysis.logic

import android.content.pm.PackageInfo
import android.support.annotation.WorkerThread
import android.util.Log
import dalvik.system.DexFile
import sk.styk.martin.apkanalyzer.model.detail.ClassPathData
import java.io.IOException
import java.util.*

/**
 * @author Martin Styk
 * @version 21.10.2017.
 */
@WorkerThread
class DexService {

    fun get(packageInfo: PackageInfo): ClassPathData {

        val packageClasses = ArrayList<String>()
        val otherClasses = ArrayList<String>()
        var innerClasses = 0

        if (packageInfo.applicationInfo != null) {

            var dexFile: DexFile? = null
            try {
                dexFile = DexFile(packageInfo.applicationInfo.sourceDir)
                val iterator = dexFile.entries()
                while (iterator.hasMoreElements()) {
                    val className = iterator.nextElement()
                    if (className != null && className.startsWith(packageInfo.applicationInfo.packageName))
                        packageClasses.add(className)
                    else
                        otherClasses.add(className)

                    if (className != null && className.contains("$")) {
                        innerClasses++
                    }
                }
            } catch (e: IOException) {
                Log.e(DexService::class.java.simpleName, e.localizedMessage)
            } finally {
                try {
                    dexFile?.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }

        return ClassPathData(
                packageClasses = packageClasses,
                otherClasses = otherClasses,
                numberOfInnerClasses = innerClasses
        )
    }
}
